File: routes/patients.js
const express = require('express');
const router = express.Router();
const patientService = require('../services/patientService');

function envelope(data, error = null) {
  return { data, error };
}

// GET /patients?last_name=&date_of_birth=&phone_number=
router.get('/', (req, res) => {
  try {
    const { last_name, date_of_birth, phone_number } = req.query;
    const patients = patientService.listPatients({ last_name, date_of_birth, phone_number });
    res.status(200).json(envelope(patients));
  } catch (err) {
    console.error('[GET /patients] error:', err);
    res.status(500).json(envelope(null, 'Internal server error'));
  }
});

// GET /patients/:id
router.get('/:id', (req, res) => {
  try {
    const patient = patientService.getPatientById(req.params.id);
    if (!patient) return res.status(404).json(envelope(null, 'Patient not found'));
    res.status(200).json(envelope(patient));
  } catch (err) {
    console.error('[GET /patients/:id] error:', err);
    res.status(500).json(envelope(null, 'Internal server error'));
  }
});

// POST /patients
router.post('/', (req, res) => {
  try {
    const patient = patientService.createPatient(req.body || {});
    res.status(201).json(envelope(patient));
  } catch (err) {
    if (err.name === 'ValidationError') {
      return res.status(422).json(envelope(null, { message: 'Validation failed', fields: err.errors }));
    }
    console.error('[POST /patients] error:', err);
    res.status(500).json(envelope(null, 'Internal server error'));
  }
});

// PUT /patients/:id
router.put('/:id', (req, res) => {
  try {
    const patient = patientService.updatePatient(req.params.id, req.body || {});
    res.status(200).json(envelope(patient));
  } catch (err) {
    if (err.name === 'NotFoundError') {
      return res.status(404).json(envelope(null, 'Patient not found'));
    }
    if (err.name === 'ValidationError') {
      return res.status(422).json(envelope(null, { message: 'Validation failed', fields: err.errors }));
    }
    console.error('[PUT /patients/:id] error:', err);
    res.status(500).json(envelope(null, 'Internal server error'));
  }
});

// DELETE /patients/:id (soft delete)
router.delete('/:id', (req, res) => {
  try {
    patientService.softDeletePatient(req.params.id);
    res.status(200).json(envelope({ deleted: true }));
  } catch (err) {
    if (err.name === 'NotFoundError') {
      return res.status(404).json(envelope(null, 'Patient not found'));
    }
    console.error('[DELETE /patients/:id] error:', err);
    res.status(500).json(envelope(null, 'Internal server error'));
  }
});

module.exports = router;

File: app/routes/patients.py
from flask import Blueprint, jsonify, request

from app import patient_service

bp = Blueprint("patients", __name__, url_prefix="/patients")


def envelope(data, error=None):
    return jsonify({"data": data, "error": error})


@bp.get("")
def list_patients():
    try:
        patients = patient_service.list_patients(
            last_name=request.args.get("last_name"),
            date_of_birth=request.args.get("date_of_birth"),
            phone_number=request.args.get("phone_number"),
        )
        return envelope(patients), 200
    except Exception as err:  # noqa: BLE001
        print("[GET /patients] error:", err)
        return envelope(None, "Internal server error"), 500


@bp.get("/<patient_id>")
def get_patient(patient_id):
    try:
        patient = patient_service.get_patient_by_id(patient_id)
        if not patient:
            return envelope(None, "Patient not found"), 404
        return envelope(patient), 200
    except Exception as err:  # noqa: BLE001
        print("[GET /patients/:id] error:", err)
        return envelope(None, "Internal server error"), 500


@bp.post("")
def create_patient():
    try:
        payload = request.get_json(silent=True) or {}
        patient = patient_service.create_patient(payload)
        return envelope(patient), 201
    except patient_service.ValidationError as err:
        return envelope(None, {"message": "Validation failed", "fields": err.errors}), 422
    except Exception as err:  # noqa: BLE001
        print("[POST /patients] error:", err)
        return envelope(None, "Internal server error"), 500


@bp.put("/<patient_id>")
def update_patient(patient_id):
    try:
        payload = request.get_json(silent=True) or {}
        patient = patient_service.update_patient(patient_id, payload)
        return envelope(patient), 200
    except patient_service.NotFoundError:
        return envelope(None, "Patient not found"), 404
    except patient_service.ValidationError as err:
        return envelope(None, {"message": "Validation failed", "fields": err.errors}), 422
    except Exception as err:  # noqa: BLE001
        print("[PUT /patients/:id] error:", err)
        return envelope(None, "Internal server error"), 500


@bp.delete("/<patient_id>")
def delete_patient(patient_id):
    try:
        patient_service.soft_delete_patient(patient_id)
        return envelope({"deleted": True}), 200
    except patient_service.NotFoundError:
        return envelope(None, "Patient not found"), 404
    except Exception as err:  # noqa: BLE001
        print("[DELETE /patients/:id] error:", err)
        return envelope(None, "Internal server error"), 500
